.github/workflows/ci.yml

```
@@ -75,156 +6,109 @@ on:
+  PART1_FILES: part1/gaussian.py part1/determinant.py part1/inverse.py part1/rank_basis.py
+  PART2_FILES: part2/decomposition.py part2/diagonalization.py part2/theme.py
+  PART3_FILES: part3/solvers.py part3/benchmark.py
+
 jobs:
+  # ── Fast gates: run first, fail before any install cost ─────────────────────
+  pre-check:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check branch name
+        if: github.event_name == 'pull_request' && github.head_ref != 'dev'
+        run: |
```

actionlint (1.7.12)

[error] "github.head_ref" is potentially untrusted. avoid using it directly in inline scripts. instead, pass it through an environment variable. see https://docs.github.com/en/actions/reference/security/secure-use#good-practices-for-mitigating-script-injection-attacks for more details

(expression)